service/history/shard/context_impl.go

```diff
@@ -1868,11 +1860,21 @@ func (s *ContextImpl) acquireShard() {
 
 		err = s.transition(contextRequestAcquired{engine: engine})
 
-		if err != nil && engine != nil {
-			// We tried to set the engine but the context was already stopped
-			engine.Stop()
+		if err != nil {
+			if engine != nil {
+				// We tried to set the engine but the context was already stopped
+				engine.Stop()
+			}
+			return err
+		}
+
+		engine, err = s.engineFuture.Get(s.lifecycleCtx)
```

Comment:
It would be nice if Future had a non-blocking Get. You can do it with an already-cancelled context, but I'm not sure it's worth it here since it is guaranteed to be set. If somehow this code changes and engineFuture is not set here, then this will block (until the shard is unloaded). So it's hardly worth checking err

You could also move the get engine part into notifyQueueProcessor?

I think it's fine any of these ways

Reply:
non-blocking Get is equal to check isReady() followed by a blocking Get() with arbitrary context value? lol

I think leaving the get engine part outside the notifyQueueProcessor util function can make the logic a bit easier, as caller knows that in this context engine is guaranteed to be set. 
If we do it in the util function which supposed to be called from anywhere, then technically, it needs to worry about the blocking issue you mentioned and maybe returns errors if engine is not ready.

Reply:
I just mean I wish the code was less verbose. You can also do
```
alreadyCancelled, cancel := context.WithCancel(context.Background())
cancel()
value, err := future.Get(alreadyCancelled)
```
I guess that's not really that bad...

Reply:
I don't have strong preference either way. But yeah the code can look a bit cleaner.

Updated the logic to use non-blocking Get and moved get engine into notifyQueueProcessor,